### ISSUE TITLE
Token cache key for Azure AD B2C users

### ISSUE THREAD

**Issue by WenningQiu** [AzureAD/microsoft-authentication-library-for-dotnet#4057]
I have noticed that the token cache key for Azure AD B2C users is a combination of the policy id and the tenant id, e.g. "-b2c_1a_signin_xxxxxdev.2d3c6047-afa5-441b-8282-314f6dfe18d4". I am afraid that the access token for different users will get mixed up in the token cache.

I have a few questions:

1. I suspect that Microsoft.Identity.Web might just pull the token cache key from the access token, as the "home_account_id" field contains the same value (see the screenshot below), but I need folks to confirm that.
2. If that's the case, is there a way to configure Azure B2C use use a better key/id?
3. Can Microsoft.Identity.Web be enhanced to compose cache keys that are unique for different B2C users?

![image](https://user-images.githubusercontent.com/4291622/229262533-024e3a11-7cad-43b1-b080-1161be37a6dd.png)

**Comment by bgavrilMS** [comment]
@WenningQiu - can you think of a scenario that would be broken? 


TLDR; the `home_account_id` is made of the `policy` and the `oid`. The `oid` is "Object ID", uniquely identifies the user. 



Both Access Tokens and Refresh Tokens are internally stored by: 

- home account id (in B2C world, this is the object ID + policy)
- environment (the host portion of the authority, i.e. xyz.b2clogin.com)
- client_id
- tenant_id / realm 
- scopes 

(the last 2 are not used for refresh tokens).

**Comment by bgavrilMS** [comment]
Not sure I understand what you mean by `oid is set by the RP`. `oid` is known by the token issuer (B2C).

Note that MSAL's force the token issuer to add the `oid` in the token response. **We do not rely on it being present in the IdToken.**


<img width="219" alt="image" src="https://user-images.githubusercontent.com/12273384/229563399-184f36f1-7cd9-49d2-a9ab-c8f4526b649e.png">
See the client_info in the picture above. This will have the `oid`. It's a simple base64 encoded json with oid and policy. Apps using MSAL do not have the option of not requesting "client info", because MSAL uses the data for token caching.

**Comment by WenningQiu** [comment]
@bgavrilMS  My RP sets `oid` in the red box below. Will that mess up `home_account_id`?
![image](https://user-images.githubusercontent.com/4291622/229567230-1d371be1-369b-4d0f-88ea-3d4d34442d18.png)

**Comment by bgavrilMS** [comment]
No, `home_account_id` is entirely driven by the token issuer (B2C) and was specially designed to help MSALs with token caching.

**Comment by WenningQiu** [comment]
"objectId" claim is part of the generated "home_account_id".

Once I make sure "objectId" claim has a unique value, the token cache key is unique for the user.
